src/inject/dynamic-theme/image.ts

```diff
@@ -50,93 +71,160 @@ async function urlToImage(url: string) {
     });
 }
 
-const MAX_ANALIZE_PIXELS_COUNT = 32 * 32;
-let canvas: HTMLCanvasElement | OffscreenCanvas;
-let context: CanvasRenderingContext2D | OffscreenCanvasRenderingContext2D;
-
-function createCanvas() {
-    const maxWidth = MAX_ANALIZE_PIXELS_COUNT;
-    const maxHeight = MAX_ANALIZE_PIXELS_COUNT;
-    canvas = document.createElement('canvas');
-    canvas.width = maxWidth;
-    canvas.height = maxHeight;
-    context = canvas.getContext('2d');
-    context.imageSmoothingEnabled = false;
+export interface ImageQueueEntry {
+    image: HTMLImageElement;
+    onResolve: (details: AnalyzeImageResults) => void;
 }
 
-function removeCanvas() {
-    canvas = null;
-    context = null;
+interface AnalyzeImageResults {
+    isDark: boolean;
+    isLight: boolean;
+    isTransparent: boolean;
+    isLarge: boolean;
 }
 
-function analyzeImage(image: HTMLImageElement) {
-    if (!canvas) {
-        createCanvas();
+// AnalyzeImage is a very slow function, so will add a manager on top of it
+// This manager will be responsible for analyzing incoming images of the queue.
+// And will only allow certain amount of images to be analyzed per second.
+// This is to avoid too much memory usage and CPU usage.
+class AnalyzeImageManager {
```

Comment:
Nice usage of asynchronous queue here! The problem is one synchronous job takes unpredictable amount of time.

But the solution is already there! [I did it 4 years ago](https://github.com/TargetProcess/tauCharts/blob/master/src/charts/task-runner.ts) :)

It ensures the tasks take ~60FPS (no more 17ms synchronously), and launches other tasks later. Since `requestAnimationFrame` is paused when the tab is hidden, it switches to `requestIdleCallback`. Not sure if it is relevant for us, because we don't want to run the job on hidden tabs and frames.

Take a look on a simpler [example](https://github.com/alexanderby/mova/blob/master/src/content/index.js#L76) I used for another extension.

I think you could improve the code by putting a TaskRunner or AsyncQueue into a separate class or class-like factory function. Maybe it can be useful in other places too.

Reply:
Done with the latest commit, I'd know a few places which would love to have this kind of `AsyncQueue`, but let's test it out for now with the images and see how it goes. Gmail is noticeably performing better on my special machine and on my PC which I didn't expect given it was already only ~5 seconds overhead, but it's now ~2-3 seconds overhead.